python/cugraph/cugraph/gnn/data_loading/bulk_sampler_io.py

```diff
@@ -54,35 +54,46 @@ def _write_samples_to_parquet(
     if partition_info != "sg" and (not isinstance(partition_info, dict)):
         raise ValueError("Invalid value of partition_info")
 
-    max_batch_id = offsets.batch_id.max()
+    # Offsets is always in order, so the last batch id is always the highest
+    max_batch_id = offsets.batch_id.iloc[len(offsets) - 1]
     results.dropna(axis=1, how="all", inplace=True)
     results["hop_id"] = results["hop_id"].astype("uint8")
 
     for p in range(0, len(offsets), batches_per_partition):
         offsets_p = offsets.iloc[p : p + batches_per_partition]
         start_batch_id = offsets_p.batch_id.iloc[0]
-        end_batch_id = offsets_p.batch_id.iloc[-1]
+        end_batch_id = offsets_p.batch_id.iloc[len(offsets_p) - 1]
+
+        reached_end = end_batch_id == max_batch_id
 
         start_ix = offsets_p.offsets.iloc[0]
-        if end_batch_id == max_batch_id:
+        if reached_end:
             end_ix = len(results)
         else:
             offsets_z = offsets[offsets.batch_id == (end_batch_id + 1)]
             end_ix = offsets_z.offsets.iloc[0]
 
-        full_output_path = os.path.join(
-            output_path, f"batch={start_batch_id}-{end_batch_id}.parquet"
-        )
         results_p = results.iloc[start_ix:end_ix].reset_index(drop=True)
 
-        results_p["batch_id"] = offsets_p.batch_id.repeat(
+        if end_batch_id - start_batch_id + 1 > len(offsets_p):
+            # This occurs when some batches returned 0 samples.
+            # To properly account this, the remaining batches are
+            # renumbered to have contiguous batch ids and the empty
+            # samples are dropped.
+            offsets_p.drop("batch_id", axis=1, inplace=True)
+            batch_id_range = cudf.Series(cupy.arange(len(offsets_p)))
+            end_batch_id = start_batch_id + len(offsets_p) - 1
```

Comment:
I think because we are renumbering `batch_ids`  here , they would no longer correspond to the  input batch_ids, this is not a problem downstream but can cause problems in the future while trying to debug failures. 

I think we should make an explicit note of this somewhere in our code because this will non intutive once we forget the implementation details.

Reply:
I think we should actually file a tech debt issue for this too , linking it in the note.

Reply:
@VibhuJawa I think this should be resolved now, please re-review